// @azure/notification-hubs/dist/commonjs/serializers/notificationDetailsSerializer.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseNotificationDetails = parseNotificationDetails;
const utils_js_1 = require("../utils/utils.js");
const core_xml_1 = require("@azure/core-xml");
/**
 * @internal
 * Parses a NotificationDetails from incoming XML.
 */
async function parseNotificationDetails(bodyText) {
    const xml = await (0, core_xml_1.parseXML)(bodyText, {
        includeRoot: true,
    });
    const notificationDetails = xml["NotificationDetails"];
    let apnsOutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["ApnsOutcomeCounts"])) {
        apnsOutcomeCounts = parseOutcomeCounts(notificationDetails["ApnsOutcomeCounts"]["Outcome"]);
    }
    let admOutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["AdmOutcomeCounts"])) {
        admOutcomeCounts = parseOutcomeCounts(notificationDetails["AdmOutcomeCounts"]["Outcome"]);
    }
    let baiduOutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["BaiduOutcomeCounts"])) {
        baiduOutcomeCounts = parseOutcomeCounts(notificationDetails["BaiduOutcomeCounts"]["Outcome"]);
    }
    let browserOutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["BrowserOutcomeCounts"])) {
        browserOutcomeCounts = parseOutcomeCounts(notificationDetails["BrowserOutcomeCounts"]["Outcome"]);
    }
    let fcmOutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["GcmOutcomeCounts"])) {
        fcmOutcomeCounts = parseOutcomeCounts(notificationDetails["GcmOutcomeCounts"]["Outcome"]);
    }
    let fcmV1OutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["FcmV1OutcomeCounts"])) {
        fcmV1OutcomeCounts = parseOutcomeCounts(notificationDetails["FcmV1OutcomeCounts"]["Outcome"]);
    }
    let xiaomiOutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["XiaomiOutcomeCounts"])) {
        xiaomiOutcomeCounts = parseOutcomeCounts(notificationDetails["XiaomiOutcomeCounts"]["Outcome"]);
    }
    let wnsOutcomeCounts;
    if ((0, utils_js_1.isDefined)(notificationDetails["WnsOutcomeCounts"])) {
        wnsOutcomeCounts = parseOutcomeCounts(notificationDetails["WnsOutcomeCounts"]["Outcome"]);
    }
    return {
        notificationId: (0, utils_js_1.getStringOrUndefined)(notificationDetails["NotificationId"]),
        location: (0, utils_js_1.getStringOrUndefined)(notificationDetails["Location"]),
        state: (0, utils_js_1.getStringOrUndefined)(notificationDetails["State"]),
        enqueueTime: (0, utils_js_1.getDateOrUndefined)(notificationDetails["EnqueueTime"]),
        startTime: (0, utils_js_1.getDateOrUndefined)(notificationDetails["StartTime"]),
        endTime: (0, utils_js_1.getDateOrUndefined)(notificationDetails["EndTime"]),
        pnsErrorDetailsUrl: (0, utils_js_1.getStringOrUndefined)(notificationDetails["PnsErrorDetailsUri"]),
        targetPlatforms: (0, utils_js_1.getStringOrUndefined)(notificationDetails["TargetPlatforms"]),
        notificationBody: (0, utils_js_1.getStringOrUndefined)(notificationDetails["NotificationBody"]),
        apnsOutcomeCounts,
        admOutcomeCounts,
        baiduOutcomeCounts,
        browserOutcomeCounts,
        fcmOutcomeCounts,
        fcmV1OutcomeCounts,
        xiaomiOutcomeCounts,
        wnsOutcomeCounts,
    };
}
function parseOutcomeCounts(counts) {
    const items = Array.isArray(counts) ? counts : [counts];
    const results = [];
    for (const item of items) {
        results.push({ state: item["Name"], count: (0, utils_js_1.getInteger)(item["Count"], "Count") });
    }
    return results;
}
